File: 149_Max Points on a Line.cpp
/**
 * Definition for a point.
 * struct Point {
 *     int x;
 *     int y;
 *     Point() : x(0), y(0) {}
 *     Point(int a, int b) : x(a), y(b) {}
 * };
 */
class Solution {
public:
    int maxPoints(vector<Point>& points) {
        int n = points.size();
        if(n==0 ||n==1) return n;
        map<double,int> xielv;
        int result = 0;
        int xielv0 = 0;
        for(int i = 0;i<n;i++)
        {
            for(int j = i+1;j<n;j++)
            {
                if(points[i].x!=points[j].x)
                {
                    double temp = (double)(points[i].y-points[j].y)/(points[i].x-points[j].x);
                    xielv[temp]++;
                    result = max(result,xielv[temp]);
                }
                else
                {
                    xielv0++;
                    result = max(result,xielv0);
                }
            }
        }
        result = sqrt(result*2)+1;
        return result;
    }
};

File: 149_Max Points on a Line.py
from fractions import Fraction
import math
class Solution:
    def getKB(self,point1,point2):
        try:
            k = Fraction(point2[1]-point1[1])/(point2[0]-point1[0])
            b = point1[1]-k*point1[0]
        except ZeroDivisionError as e:
            k = None
            b = point1[0]
        return k,b
    def maxPoints(self, points) -> int:
        points.sort(key=lambda x: x[0])
        #print(points)
        if len(points)<2:
            return len(points)
        d = {}
        for i in range(len(points)):
            for j in range(i+1,len(points)):
                kb = self.getKB(points[i],points[j])
                key = str(kb)
                keyi = str(points[i])
                keyj = str(points[j])
                if key not in d:
                    d[key] = {keyi,keyj}
                else:
                    d[key].add(keyi)
                    d[key].add(keyj)
        #print(d)
        count = {}
        for key in d:
            count[key] = 0
        result = 0
        for i in range(len(points)):
            keyi = str(points[i])
            for key in d:
                if keyi in d[key]:
                    count[key]+=1
                    if count[key]> result:
                        result = count[key]
        return result
